bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookKeeperAdmin.java

```diff
@@ -882,27 +941,74 @@ private void asyncRecoverLedgerFragment(final LedgerHandle lh,
      *            - ledgerHandle
      * @param ledgerFragment
      *            - LedgerFragment to replicate
-     * @param targetBookieAddress
-     *            - target Bookie, to where entries should be replicated.
      */
     public void replicateLedgerFragment(LedgerHandle lh,
+            final LedgerFragment ledgerFragment)
+            throws InterruptedException, BKException {
+        Optional<Set<BookieSocketAddress>> excludedBookies = Optional.empty();
+        Map<Integer, BookieSocketAddress> targetBookieAddresses =
+                getReplacedBookiesByIndexes(lh, ledgerFragment.getEnsemble(),
+                        ledgerFragment.getBookiesIndexes(), excludedBookies);
+        replicateLedgerFragment(lh, ledgerFragment, targetBookieAddresses);
+    }
+
+    private void replicateLedgerFragment(LedgerHandle lh,
             final LedgerFragment ledgerFragment,
-            final BookieSocketAddress targetBookieAddress)
+            final Map<Integer, BookieSocketAddress> targetBookieAddresses)
```

Comment:
why are there multiple bookies being passed for a single fragment? A fragment is the subdivision of a segment that is stored on a single bookie.

Reply:
Ah, you're changing the definition of fragment later on too.

Reply:
yes, a fragment is a range of ledger entries. it is not per bookie. so when we replicate a fragment, it can be more efficient on replicating to multiple bookies.